### PR TITLE
8301121: RichTextArea Control (Incubator)

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Incubating a new feature - rich text control, RichTextArea, intended to bridge the functional gap with Swing and its StyledEditorKit/JEditorPane.  The main design goal is to provide a control that is complete enough to be useful out-of-the box, as well as open to extension by the application developers.

This is a complex feature with a large API surface that would be nearly impossible to get right the first time, even after an extensive review.  We are, therefore, introducing this in an incubating module, javafx.incubator.controls.   This will allow us to evolve the API in future releases without the strict compatibility constraints that other JavaFX modules have.

Please check out two manual test applications - one for RichTextArea (RichTextAreaDemoApp) and one for the CodeArea (CodeAreaDemoApp).  Also, a small example provides a standalone rich text editor, see RichEditorDemoApp.

References

[0] Proposal: https://github.com/andy-goryachev-oracle/Test/blob/rich.jep.review/doc/RichTextArea/RichTextArea.md
[1] Discussion points: https://github.com/andy-goryachev-oracle/Test/blob/rich.jep.review/doc/RichTextArea/RichTextAreaDiscussion.md
[2] API specification (javadoc): https://cr.openjdk.org/~angorya/RichTextArea/javadoc
[3] Draft Pull Request for API comments and feedback: https://github.com/openjdk/jfx/pull/1374
[4] RichTextArea RFE: https://bugs.openjdk.org/browse/JDK-8301121
[5] Missing APIs related to rich text control: https://bugs.openjdk.org/browse/JDK-8300569

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warnings
&nbsp;⚠️ Patch contains a binary file (modules/jfx.incubator.richtext/src/main/docs/jfx/incubator/scene/control/rich/doc-files/CodeArea.png)
&nbsp;⚠️ Patch contains a binary file (modules/jfx.incubator.richtext/src/main/docs/jfx/incubator/scene/control/rich/doc-files/RichTextArea.png)
&nbsp;⚠️ Patch contains a binary file (tests/manual/RichTextAreaDemo/src/com/oracle/demo/richtext/rta/animated.gif)

### Issue
 * [JDK-8301121](https://bugs.openjdk.org/browse/JDK-8301121): RichTextArea Control (Incubator) (**Enhancement** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1374/head:pull/1374` \
`$ git checkout pull/1374`

Update a local copy of the PR: \
`$ git checkout pull/1374` \
`$ git pull https://git.openjdk.org/jfx.git pull/1374/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1374`

View PR using the GUI difftool: \
`$ git pr show -t 1374`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1374.diff">https://git.openjdk.org/jfx/pull/1374.diff</a>

</details>

### CODE REVIEW

**Comment by karthikpandelu** [comment]
I tried the manual test applications and have few observations:
**RichTextAreaDemoApp:**

1. Is this app meant to be non editable? I see that the `editable` checkbox is not changing anything.
2. There is a `Test` button in the options pane at the bottom. I couldn't really figure out its usage.
3. In Mac, horizontal scrolling is not working using the trackpad whereas vertical scrolling is working as expected. This is in all the test apps.

**CodeAreaDemoApp:**

1. Selecting the line number checkbox throws NPE and code area becomes blank. Deselecting the checkbox brings back the content.
2. Changing font size is not changing text size.
I observed these issue when I ran the application for the first time. When I relaunched the application, I couldn't reproduce these issues.

In RichEditorDemoApp, bold, underline and italic buttons did not work when the app was launched for the first time. After relaunching the app, only italic button is not working.

**Comment by andy-goryachev-oracle** [comment]
Thank you @karthikpandelu for testing!

1. some of the issues you've seen with exceptions being thrown are fixed in the later code, I did not want to update in the middle of your review (will update shortly)
2. the `editable` checkbox modifies the property in the control, you'd need to have an editable model to be able to edit.  the purpose of this property is to disable editing even if the model itself is editable.
3. `Test` button appends some text via control's APIs, will fix the button.
4. The horizontal scrolling is not implemented - totally missed it.  Thank you for uncovering this requirements gap!

**Review by aghaisas — COMMENTED** [review]
I started taking a look.  A few general questions -

1) What level of CSS support does the RichTextArea control have?
For e.g. TextArea has this level of support -
https://docs.oracle.com/javafx/2/api/javafx/scene/doc-files/cssref.html#textarea

2) How is 2B row limit imposed?

3) How is the performance impact as and when model size increases? As it is a sliding window design, I believe, it should become constant at some point of model size.

**Comment by andy-goryachev-oracle** [comment]
> 1. What level of CSS support does the RichTextArea control have?
>   For e.g. TextArea has this level of support -
>   https://docs.oracle.com/javafx/2/api/javafx/scene/doc-files/cssref.html#textarea

A good question.

At the moment, `RichTextArea` exposes only two properties via CSS:
```
-fx-content-padding
-fx-wrap-text
```

While `CodeArea` adds three more:
```
-fx-font
-fx-line-spacing
-fx-tab-size
```

We could add more.  The question is - does it make sense to make all the trivial properties styleable?  What is the general rule that we should use to make properties styleable?


> 2. How is 2B row limit imposed?

Paragraph index is an `int`.


> 3. How is the performance impact as and when model size increases? As it is a sliding window design, I believe, it should become constant at some point of model size.

Correct - for most operations, the performance is a function of the text being laid out in the sliding window.  Obviously, very long paragraphs will slow things down.  (There is currently a bug where editing the Writing Systems model is slow, I'll be working on that shortly).

Of course, exporting large chunks of text from a very large model will take time proportional to the size of the text being exported.  Copying a large selection might throw an OOM Error.

One thing was explicitly added to the layout logic: when `useContentHeight` is enabled, the layout stops at some arbitrary height so as not to lock up the UI, see `Params.MAX_HEIGHT_SAFEGUARD`.

File: modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/rich/model/StyleAttrs.java [hunk 1]
```diff
@@ -0,0 +1,720 @@
...
+import javafx.scene.paint.Paint;
+import javafx.scene.text.Font;
+import javafx.scene.text.Text;
+import javafx.scene.text.TextAlignment;
+import com.sun.jfx.incubator.scene.control.rich.CssStyles;
+import com.sun.jfx.incubator.scene.control.rich.util.RichUtils;
+
+/**
+ * An immutable object containing style attributes.
+ * TODO name: StyleSet? (though it's not a set)
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/rich/model/StyleAttrs.java:44]: What would you suggest?
Place in different classes, re-arrange in two blocks, or something else?

File: modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/rich/CodeArea.java [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        CodeTextModel m = codeModel();
+        if (m != null) {
+            m.setDecorator(d);
+        }
+    }
+
+    /**
+     * This convenience method returns the syntax decorator value in the model,
+     * or null if the said model is null.
+     * @return the syntax devocrator value, or null
```
Comment (karthikpandelu) [inline modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/rich/CodeArea.java:116]: Typo in decorator

**Review by karthikpandelu — COMMENTED** [review]
Added few minor inline comments. Meanwhile I will be continuing the review.

File: modules/javafx.base/src/main/java/com/sun/javafx/ModuleUtil.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+            var stackWalker = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
+            var callerClass = stackWalker.walk(s ->
+                s.dropWhile(f -> {
+                    var clazz = f.getDeclaringClass();
+                    return ModuleUtil.class.equals(clazz) || MODULE_JAVA_BASE.equals(clazz.getModule());
+                })
+                .map(StackWalker.StackFrame::getDeclaringClass)
+                .findFirst()
+                .orElseThrow(IllegalStateException::new));
+            //System.err.println("callerClass = " + callerClass);
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.base/src/main/java/com/sun/javafx/ModuleUtil.java:54]: @kevinrushforth I think this comment also applies to your next incubator module PR...

File: doc-files/behavior/RichTextAreaBehavior.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+|MOVE_PARAGRAPH_DOWN       |Moves the caret to the end of the current paragraph, or, if already there, to the end of the next paragraph
+|MOVE_PARAGRAPH_UP         |Moves the caret to the start of the current paragraph, or, if already there, to the start of the previous paragraph
+|MOVE_RIGHT                |Moves the caret one symbol to the right
+|MOVE_TO_DOCUMENT_END      |Moves the caret to after the last character of the text
+|MOVE_TO_DOCUMENT_START    |Moves the caret to before the first character of the text
+|MOVE_TO_PARAGRAPH_END     |Moves the caret to the end of the paragraph at caret
+|MOVE_TO_PARAGRAPH_START   |Moves the caret to the beginning of the paragraph at caret
+|MOVE_UP                   |Moves the caret one visual text line up
+|MOVE_WORD_LEFT            |Moves the caret one word left (previous word if LTR, next word if RTL)
+|MOVE_WORD_NEXT            |Moves the caret to the beginning of the next word, or next paragraph if at the start of an empty paragraph
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply doc-files/behavior/RichTextAreaBehavior.md:32]: added _DEL

**Comment by andy-goryachev-oracle** [comment]
@aghaisas @kevinrushforth please check out the updated public api and javadoc

https://cr.openjdk.org/~angorya/RichTextArea/javadoc/

File: modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/rich/skin/RichTextAreaSkin.java [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    }
+
+    /**
+     * Pastes the clipboard content at the caret, or, if selection exists, replacing the selected text.
+     * The format must be supported by the model, and the skin must be installed,
+     * otherwise this method has no effect.
+     *
+     * @param format data format
+     */
+    public void paste(DataFormat format) {
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/rich/skin/RichTextAreaSkin.java:270]: thank you, will address all suggestions.

**Comment by andy-goryachev-oracle** [comment]
Closing this internal review PR in favor of #1524

**Pull request closed without merging**